src/Service/ForecastReportService.php

```diff
@@ -0,0 +1,184 @@
+
+        // Create an new instance of ForecastReportData
+        $forecastReportData = new ForecastReportData();
+
+        // Loop through each worklog
+        foreach ($invoiceAttachedWorklogs as $worklog) {
+            $projectId = $worklog->getProject()->getId();
+
+            if (!$projectId) {
+                throw new \Exception('Project id is null');
+            }
+            $projectName = $worklog->getProject()?->getName() ?? '[no project name]';
+
+            // If the project isn't already in the forecast, add it
+            if (!isset($forecastReportData->projects[$projectId])) {
+                $forecastReportData->projects[$projectId] = new ForecastReportProjectData($projectId);
+            }
+
+            if (is_object($forecastReportData->projects[$projectId])) {
+                $forecastReportData->projects[$projectId]->projectName = $projectName;
+            }
+
+            // Get current project from forecast
+            $currentProject = $forecastReportData->projects[$projectId];
+
+            if (!$currentProject) {
+                throw new \Exception('Project instance was not found');
+            }
+
+            // Calculate worklog time in hours
+            $worklogTime = ($worklog->getTimeSpentSeconds() / 3600);
+
+            // Check if worklog is billed
+            $isWorklogBilled = $worklog->isBilled();
+
+            // Tally up total project hours based on whether the worklog is billed
+            $currentProject->invoiced += $worklogTime;
+            if ($isWorklogBilled) {
+                $currentProject->invoicedAndRecorded += $worklogTime;
+            }
+
+            // Get issue details from worklog
+            $issueId = $worklog->getIssue()->getProjectTrackerKey();
+            $issueLink = $worklog->getIssue()->getLinkToIssue();
+            $issueTag = $worklog->getIssue()->getEpicName() ?: '[no tag]';
+
+            // Add issue in the project if it does not exist
+            if (!isset($currentProject->issues[$issueTag])) {
+                $currentProject->issues[$issueTag] = new ForecastReportIssueData($issueTag);
+                $currentProject->issues[$issueTag]->issueId = $issueId;
+                $currentProject->issues[$issueTag]->issueLink = $issueLink;
+            }
+
+            // Get current issue from project
+            $currentIssue = $currentProject->issues[$issueTag];
+
+            // Add up the invoiced hours to the current issue
+            $currentIssue->invoiced += $worklogTime;
+            if ($isWorklogBilled) {
+                $currentIssue->invoicedAndRecorded += $worklogTime;
+            }
+
+            // Get version details from issue
+            $issueVersions = $worklog->getIssue()->getVersions();
+            $issueVersion = count($issueVersions) > 0 ? implode('', array_map(function ($version) { return $version->getName(); }, $issueVersions->toArray())) : '[no version]';
```

Comment:
shouldn't the versions string be separated with ', ' instead of '' ?

Reply:
You are right. With Leantime as the data source, issues can only contain a single version, so i didn't think any of it.

 I changed this.